### PR TITLE
 Allow executing a built in function against the root of a document 

### PR DESCRIPTION
**Pull request by EricBorczuk**
**What this PR does**:
Allow executing a built in function against the root of a document (before it was producing 405 Method Not Allowed because it was matching a different route)

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [x] Documentation added/updated
- [x] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java [hunk 1]
```diff
@@ -113,7 +113,7 @@ public class BuiltInFunctionResource {
       })
   @POST
   @Path(
-      "{collection:\\w+}/{document-id}{slash: /?}{document-path: (.*)?}/function{trailingSlash: /?}")
+      "{collection:\\w+}/{document-id: [^/]+}{slash: /?}{document-path: (.+)?}/function{trailingSlash: /?}")
```
Comment (tatu-at-datastax) [inline apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java:116]: On `document-id`: would slash actually be allowed otherwise? I assumed it usually wouldn't. But I assume that it would be (otherwise wouldn't have need to prevent it)

File: apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java [hunk 2]
```diff
@@ -124,8 +124,12 @@ public Uni<RestResponse<Object>> executeBuiltInFunction(
       @QueryParam("raw") boolean raw,
       @NotNull(message = "payload not provided") @Valid BuiltInFunctionDto body) {
     ExecutionContext context = ExecutionContext.create(profile);
+
     List<String> subPath =
-        documentPath.stream().map(PathSegment::getPath).collect(Collectors.toList());
+        documentPath.stream()
+            .map(PathSegment::getPath)
+            .filter(p -> !p.isEmpty())
```
Comment (tatu-at-datastax) [inline apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java:131]: So empty path segments are allowed, just trimmed out?

File: apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentDeleteResource.java [hunk 3]
```diff
@@ -158,7 +158,10 @@ public Uni<RestResponse<Object>> deleteSubDocument(
       @QueryParam("profile") boolean profile) {
     ExecutionContext context = ExecutionContext.create(profile);
     List<String> subPath =
-        documentPath.stream().map(PathSegment::getPath).collect(Collectors.toList());
+        documentPath.stream()
+            .map(PathSegment::getPath)
+            .filter(p -> !p.isEmpty())
+            .collect(Collectors.toList());
```
Comment (tatu-at-datastax) [inline apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentDeleteResource.java:164]: Not a big deal but since it's called from a few resources, could also extract into helper method.

File: apis/sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResourceIntegrationTest.java [hunk 4]
```diff
@@ -411,11 +411,9 @@ public void setUpdateManyValuesBaseDoc() {
           .get(BASE_PATH, DEFAULT_NAMESPACE, DEFAULT_COLLECTION, DOCUMENT_ID)
           .then()
           .body("documentId", equalTo(DOCUMENT_ID))
-          .body("data", jsonPartEquals("b[1].nested", "value"))
```
Comment (tatu-at-datastax) [inline apis/sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResourceIntegrationTest.java]: Why were these removed?

**Review by tatu-at-datastax — APPROVED** [review]

File: apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java (hunk 2 above)
Comment (EricBorczuk) [reply apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java:131]: Yeah, if you do something like /api/v2/namespaces/x/collections/y/document/a//b/c, I just figured it should ignore the typo

File: apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java (hunk 1 above)
Comment (EricBorczuk) [reply apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java:116]: Yea I noticed that because of how quarkus URI parsing works with slashes, it actually slurps everything between literal `/` for each token. So I was getting matches on `{document-id: .+}{slash: /?}{document-path: (.+)?}` that were actually matching `abcde/fghij` as the value of document-id (which is ...not right)

File: apis/sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResourceIntegrationTest.java (hunk 4 above)
Comment (EricBorczuk) [reply apis/sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResourceIntegrationTest.java]: This entire test fn was added - but then I realized that because it is updating data on the root of the document, the data at the root is just different.

In the test this was copied from, I was setting data on a nested path, and then checking that both the data at the nested path previously and the new data were both present.

Comment (tatu-at-datastax) [reply apis/sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResourceIntegrationTest.java]: Ok

File: apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java (hunk 1 above)
Comment (tatu-at-datastax) [reply apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java:116]: Yeah, just odd how it would include slash within if you do NOT include pattern. I would expect it to essentially use "anything but slash". But I guess not :)

File: apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java (hunk 2 above)
Comment (EricBorczuk) [reply apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java:131]: Yeaaaa that actually might be a problem in certain cases...it would mean that we have to restrict JSON fields from having empty string as well, or else you could write JSON with an empty string as field and then never be able to read that path directly via the URI. Let me check to see how we handle this, because also I think that empty-string serves as the delimiter for where the nesting ends in the cassandra row

Comment (tatu-at-datastax) [reply apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java:131]: Right. To me it seems reasonable not to allow these, but I am not 100% sure how Ivan feels about that.
I could see empty String as key for some sort of special metadata, and it is probably used by someone somewhere.
Does not mean we have to support it but we need to decide one way or the other.

**Review by ivansenic — CHANGES_REQUESTED** [review]
Few comments on my side:

1. was a trailing slash really necessary? I would say no, why was this added?
2. the whole change of the endpoints, now require an update of the path matching for the micrometer metrics.. however, with being so complex now, I have no idea how to adapt this now..

I am really opting in only fixing the `/function` stuff.. Blocking merge until we resolve this..

**Comment by EricBorczuk** [comment]
The trailing-slash changes were done because Val's mongoose stuff allows either, and it made sense for this to support both if it was easy to do...but if micrometer metrics become impossible, then maybe I would cut it...

**Comment by jeffreyscarpenter** [comment]
Can we rename this PR to be more informative, maybe something like "URI format fixes to enable stargate-mongoose"?

**Comment by EricBorczuk** [comment]
I actually removed the trailing slash changes from this PR, so it only has one significant change. Updated the description to reflect that

File: apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java [hunk 5]
```diff
@@ -112,7 +112,7 @@ public class BuiltInFunctionResource {
         @APIResponse(ref = OpenApiConstants.Responses.GENERAL_503),
       })
   @POST
-  @Path("{collection:\\w+}/{document-id}/{document-path:.*}/function")
+  @Path("{collection:\\w+}/{document-id: [^/]+}{slash: /?}{document-path: (.+)?}/function")
```
Comment (ivansenic) [inline apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java:115]: OK, but we still have to fix the micrometer template on this line:

https://github.com/stargate/stargate/blob/2384046a394613eb2d9a6d4dcea2e7712eea4f5f/apis/sgv2-docsapi/src/main/resources/application.yaml#L111

Seems like only the right hand side has to be updated to the new path spec from that `@Path`.. The best way to test this is to hit that endpoint without authentication and see what's get counted in the `/metrics`.. And then hit with the token and assure both calls are having same path tag..

Failing to do this we are risking of a metrics explosion, it's very high risk imo.. Please let me know if you have questions!

**Review by ivansenic — CHANGES_REQUESTED** [review]

File: apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java (hunk 5 above)
Comment (EricBorczuk) [reply apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java:115]: I feel like the left side needs to be changed, too - no? Because `/v2/namespaces/X/collections/Y/documentId/function` doesn't match the left side regex?

**Review by ivansenic — COMMENTED** [review]
OK the metric changes are fine, tested and it works :heavy_check_mark: 

However, have a look on the openapi, the endpoint is displayed as:

```
[/v2/namespaces/{namespace}/collections/{collection}/{document-id}{slash}{document-path}/function](http://localhost:8180/swagger-ui/#/Documents/post_v2_namespaces__namespace__collections__collection___document_id__slash__document_path__function)
Executes a a built-in function
```

And built in curl is:

```
curl -X 'POST' \
  'http://localhost:8180/v2/namespaces/cycling/collections/events/id{slash}path/function' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "operation": "$pop"
}'
```

Is this actually the right way to develop this? Should we just simply add a new endpoint with 

```
{collection:\\w+}/{document-id}/function
```

and delegate to that method below.. I am completely confused with the slash stuff? Is slash default when targeting the root document?

File: apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java [hunk 6]
```diff
@@ -112,9 +167,8 @@ public class BuiltInFunctionResource {
         @APIResponse(ref = OpenApiConstants.Responses.GENERAL_503),
       })
   @POST
-  @Path("{collection:\\w+}/{document-id}/{document-path:.*}/function")
+  @Path("{collection:\\w+}/{document-id}/{document-path: .+}/function")
```
Comment (ivansenic) [inline apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java:170]: why this change?

File: apis/sgv2-docsapi/src/main/resources/application.yaml [hunk 7]
```diff
@@ -108,7 +108,7 @@ quarkus:
           /v2/namespaces/[\\w]+/collections/[\\w]+/batch=/v2/namespaces/{namespace:\\w+}/collections/{collection:\\w+}/batch,
           /v2/namespaces/[\\w]+/collections/[\\w]+/json-schema=/v2/namespaces/{namespace:\\w+}/collections/{collection:\\w+}/json-schema,
           /v2/namespaces/[\\w]+/collections/[\\w]+/upgrade=/v2/namespaces/{namespace:\\w+}/collections/{collection:\\w+}/upgrade,
-          /v2/namespaces/[\\w]+/collections/[\\w]+/(.*?)/.*/function=/v2/namespaces/{namespace:\\w+}/collections/{collection:\\w+}/{document-id}/{document-path:.*}/function,
+          /v2/namespaces/[\\w]+/collections/[\\w]+/(.*?)/?.*/function=/v2/namespaces/{namespace:\\w+}/collections/{collection:\w+}/{document-id: [^/]+}{slash: /?}{document-path: (.+)?}/function,
```
Comment (ivansenic) [inline apis/sgv2-docsapi/src/main/resources/application.yaml:111]: This is not valid anymore:

1. new endpoint added
2. other endpoint has changed `.*` to `.+`

Not sure if this is correct, but should be:

```
/v2/namespaces/[\\w]+/collections/[\\w]+/(.*?)/.*/function=/v2/namespaces/{namespace:\\w+}/collections/{collection:\\w+}/{document-id}/{document-path:.+}/function
/v2/namespaces/[\\w]+/collections/[\\w]+/(.*?)/function=/v2/namespaces/{namespace:\\w+}/collections/{collection:\\w+}/{document-id}/function
```

File: apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java [hunk 8]
```diff
@@ -123,8 +177,12 @@ public Uni<RestResponse<Object>> executeBuiltInFunction(
       @QueryParam("raw") boolean raw,
       @NotNull(message = "payload not provided") @Valid BuiltInFunctionDto body) {
     ExecutionContext context = ExecutionContext.create(profile);
+
     List<String> subPath =
-        documentPath.stream().map(PathSegment::getPath).collect(Collectors.toList());
+        documentPath.stream()
+            .map(PathSegment::getPath)
+            .filter(p -> !p.isEmpty())
+            .collect(Collectors.toList());
```
Comment (ivansenic) [inline apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java:185]: can this also be reverted now to the old way or now? not sure..

**Review by ivansenic — APPROVED** [review]
Approved, but there are still some issues..

File: apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java (hunk 6 above)
Comment (EricBorczuk) [reply apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java:170]: accidental! I will revert

File: apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java (hunk 8 above)
Comment (EricBorczuk) [reply apis/sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/BuiltInFunctionResource.java:185]: Yeah, it can - shouldn't make a difference now

**Review by ivansenic — APPROVED** [review]

**Comment by tatu-at-datastax** [comment]
@EricBorczuk Could you merge this now? We are planning on getting this in 2.0.0-BETA-4 build.
